libraries/rush-lib/src/logic/cobuild/CobuildLock.ts

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { InternalError } from '@rushstack/node-core-library';
+
+import type { CobuildConfiguration } from '../../api/CobuildConfiguration';
+import type { OperationStatus } from '../operations/OperationStatus';
+import type { ICobuildContext } from './ICobuildLockProvider';
+import type { ProjectBuildCache } from '../buildCache/ProjectBuildCache';
+
+const KEY_SEPARATOR: ':' = ':';
+
+export interface ICobuildLockOptions {
+  /**
+   * {@inheritdoc CobuildConfiguration}
+   */
+  cobuildConfiguration: CobuildConfiguration;
+  /**
+   * {@inheritdoc ICobuildContext.clusterId}
+   */
+  cobuildClusterId: string;
+  /**
+   * {@inheritdoc ICobuildContext.packageName}
+   */
+  packageName: string;
+  /**
+   * {@inheritdoc ICobuildContext.phaseName}
+   */
+  phaseName: string;
+  projectBuildCache: ProjectBuildCache;
+  /**
+   * The expire time of the lock in seconds.
+   */
+  lockExpireTimeInSeconds: number;
+}
+
+export interface ICobuildCompletedState {
+  status: OperationStatus.Success | OperationStatus.SuccessWithWarning | OperationStatus.Failure;
+  cacheId: string;
+}
+
+export class CobuildLock {
+  public readonly cobuildConfiguration: CobuildConfiguration;
+  public readonly projectBuildCache: ProjectBuildCache;
+
+  private _cobuildContext: ICobuildContext;
+
+  public constructor(options: ICobuildLockOptions) {
+    const {
+      cobuildConfiguration,
+      projectBuildCache,
+      cobuildClusterId: clusterId,
+      lockExpireTimeInSeconds,
+      packageName,
+      phaseName
+    } = options;
+    const { cobuildContextId: contextId, cobuildRunnerId: runnerId } = cobuildConfiguration;
+    const { cacheId } = projectBuildCache;
+    this.cobuildConfiguration = cobuildConfiguration;
+    this.projectBuildCache = projectBuildCache;
+
+    if (!cacheId) {
+      // This should never happen
+      throw new InternalError(`Cache id is require for cobuild lock`);
+    }
+
+    if (!contextId) {
+      // This should never happen
+      throw new InternalError(`Cobuild context id is require for cobuild lock`);
+    }
```

Comment:
Will this ever happen?

Reply:
In the `CobuildConfiguration`, the `cobuildEnabled` will be force set to `false` when missing context id. Then, the code path can never run into `CobuildLock`. 

Does it sound good to you?